objects/SetupChecker/script.lua

```diff
@@ -1295,3 +1221,84 @@ function tFind(table, needle)
     end
     return nil
 end
+---
+
+--- Update the UI XML of the current object.
+-- This takes a list of function to use to update the XML table.
+-- Each function should take a table element, update it with any relevant changes
+-- and return `true` if updateXml should recurse into the children of the element.
+-- Note: The functions should be prepared to be called on child elements, even if
+-- recursion was not requested.
+-- @param updateFunctions The list of functions to use to update the table
+function updateXml(updateFunctions)
+    local function recurse(t)
+        local shouldRecurse = false
+        for _, f in pairs(updateFunctions) do
+            if f(t) then
+                shouldRecurse = true
+            end
+        end
+        if shouldRecurse then
+            for _, v in pairs(t.children) do
+                recurse(v)
+            end
+        end
+    end
+    local t = self.UI.getXmlTable()
+    for _,v in pairs(t) do
+        recurse(v)
+    end
+    self.UI.setXmlTable(t, {})
+end
+--- Apply an XML UI update function on an element with a specific id
+-- This returns a function suitable to pass to `updateXml` that applies the
+-- given function to the element with the given id. All the parent elements of the
+-- desired element must have the id present in the `recurse` attribute for this to
+-- find the element.
+-- @param id The id of the element to update
+-- @param f The function that updates the element
+function matchRecurse(id, f)
+    return function (t)
+        if t.attributes.recurse ~= nil and string.match(t.attributes.recurse, id) then
+            if t.attributes.id == id then
+                f(t)
+            else
+                return true
+            end
+        end
```

Comment:
so presumably there is an implied return nil for this function, but should probably be more explicit in nature